Jenkinsfile

```diff
@@ -1,117 +1,151 @@
-                }
-            }
+pipeline {
+  agent {
+    label 'centos-latest'
+  }
 
-            wrap([$class: 'Xvnc', useXauthority: true]) {
-                stage('test with Eclipse 2023-09 and Xtext nighly') {
-                    try{
-                        sh "JAVA_HOME=${java17Home} ${mvnHome}/bin/mvn -P!xtext-examples -Dtycho-version=2.7.5 -Dtarget-platform=eclipse_2023_09-xtext_nightly ${mvnParams} clean integration-test"
-                    }finally {
-                        junit '**/TEST-*.xml'
-                    }
-                }
-            }
+  triggers {
+     cron('H 23 * * *')
+  }
 
-            if(env.BRANCH_NAME?.toLowerCase() == 'master' ||
-               env.BRANCH_NAME?.toLowerCase()?.startsWith('release_')) {
-                stage('deploy') {
-                    withCredentials([file(credentialsId: 'secret-subkeys.asc', variable: 'KEYRING')]) {
-                        sh '''
-                        rm -r xpect-local-maven-repository
-                        gpg --batch --import "${KEYRING}"
-                        for fpr in $(gpg --list-keys --with-colons  | awk -F: '/fpr:/ {print $10}' | sort -u);
-                        do
-                            echo -e "5\ny\n" | gpg --batch --command-fd 0 --expert --edit-key $fpr trust;
-                        done
-                        '''
-                        sh "${mvnHome}/bin/mvn -P!tests -P!xtext-examples -P maven-publish -Dtarget-platform=eclipse_2023_03-xtext_2_31_0 ${mvnParams} clean deploy"
-                    }
-                }               
-            }
+  options {
+    buildDiscarder(logRotator(numToKeepStr: '10'))
+    disableConcurrentBuilds()
+    skipDefaultCheckout false
+  }
+
+  tools {
+    maven 'apache-maven-latest'
+    jdk 'temurin-jdk17-latest'
+  }
+
+  environment {
+    PUBLISH_LOCATION = 'updates'
+    BUILD_TIMESTAMP = sh(returnStdout: true, script: 'date +%Y%m%d%H%M').trim()
+  }
+
+  parameters {
+    choice(
+      name: 'BUILD_TYPE',
+      choices: ['nightly', 'milestone', 'release'],
+      description: '''
+        Choose the type of build.
+        Note that a release build will not promote the build, but rather will promote the most recent milestone build.
+        '''
+    )
+
+    choice(
+      name: 'TARGET_PLATFORM',
+      choices: [ 'eclipse_2023_03-xtext_2_31_0', 'eclipse_2023_09-xtext_nightly' ],
+      description: '''
+        Choose the named target platform against which to compile and test.
+        This is relevant only for nightly and milestone builds.
+        '''
+    )
+
+    booleanParam(
+      name: 'ECLIPSE_SIGN',
+      defaultValue: true,
+      description: '''
+        Choose whether or not the bundles will be signed.
+        This is relevant only for nightly and milestone builds.
+      '''
+    )
+
+    booleanParam(
+      name: 'PROMOTE',
+      defaultValue: true,
```

Comment:
do we want this on all branches?

Reply:
If you look at the later logic, if only allows the master branch to promote.

Reply:
This part ensure that only master can produce signed content (so no user can create a PR with changes that are signed) and only when signing is enabled, only on the master branch is the promotion parameter respected.
```
          if (env.BRANCH_NAME == 'master') {
            // Only sign the master branch.
            env.ECLIPSE_SIGN = params.ECLIPSE_SIGN
          } else {
            env.ECLIPSE_SIGN =  false
          }

            // Only promote signed builds.
          env.PROMOTE = params.PROMOTE && (env.ECLIPSE_SIGN == 'true')
```

Reply:
i see.

Reply:
i wonder if we need to add a target platform = older one as well.
as xtext in contrast to emf cannot generate against old runtimes
(no issue as of now as no changes between Xtext versions currently)

Reply:
These are the only two defined in the repository:
```
    choice(
      name: 'TARGET_PLATFORM',
      choices: [ 'eclipse_2023_03-xtext_2_31_0', 'eclipse_2023_09-xtext_nightly' ],
```
Maybe we don't want a choice but rather build them both (or all if there is more than one).  Though I wonder what will happen if we start using new bundle names...

Reply:
Note that for EMF I just trigger my build with different target platform choices:

https://ci.eclipse.org/emf/job/all-target-platforms/

But I only do that occasionally...

Certainly there are different design options here....

Reply:
the question is how actively people keep an eye.
> Maybe we don't want a choice but rather build them both (or all if there is more than one). Though I wonder what will happen if we start using new bundle names...

i assume for these we might need two options deps or so.
but i did not look yet how actively xpect uses classgraph

Reply:
and yes right now we run both.
but deploy only the result of the first target platform